// es-dev-server/dist/plugins/babelTransformPlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const compatibility_transform_1 = require("../utils/compatibility-transform");
const user_agent_compat_1 = require("../utils/user-agent-compat");
const message_channel_1 = require("../utils/message-channel");
const strip_ansi_1 = tslib_1.__importDefault(require("strip-ansi"));
const dom5_fork_1 = require("@open-wc/building-utils/dom5-fork");
const building_utils_1 = require("@open-wc/building-utils");
const parse5_1 = require("parse5");
const url_1 = require("url");
const utils_1 = require("../utils/utils");
function createFilePath(context, rootDir) {
    return url_1.fileURLToPath(new url_1.URL(`.${context.path}`, `${url_1.pathToFileURL(rootDir)}/`));
}
function babelTransformPlugin(config) {
    const compatibilityTransform = compatibility_transform_1.createCompatibilityTransform(config);
    const { rootDir } = config;
    async function transformJs(context, code) {
        const filePath = createFilePath(context, rootDir);
        const transformModule = context.URL.searchParams.has('transform-systemjs');
        const uaCompat = user_agent_compat_1.getUserAgentCompat(context);
        return compatibilityTransform({
            uaCompat,
            filePath,
            code,
            transformModule,
        });
    }
    return {
        async transform(context) {
            // transform a single file
            if (context.response.is('js') && !utils_1.isPolyfill(context.url)) {
                try {
                    return { body: await transformJs(context, context.body) };
                }
                catch (error) {
                    const filePath = createFilePath(context, rootDir);
                    let errorMessage = `Error compiling: ${error.message}`;
                    if (errorMessage.startsWith(filePath)) {
                        errorMessage = errorMessage.replace(filePath, context.url);
                    }
                    // send compile error to browser for logging
                    context.body = errorMessage;
                    context.status = 500;
                    message_channel_1.sendMessageToActiveBrowsers('error-message', JSON.stringify(strip_ansi_1.default(errorMessage)));
                    console.error(`\n${errorMessage}`);
                }
            }
            // transform inline JS
            if (context.response.is('html')) {
                const documentAst = parse5_1.parse(context.body);
                const scriptNodes = building_utils_1.findJsScripts(documentAst, {
                    jsScripts: true,
                    jsModules: true,
                });
                for (const node of scriptNodes) {
                    const code = dom5_fork_1.getTextContent(node);
                    const resolvedCode = await transformJs(context, code);
                    dom5_fork_1.setTextContent(node, resolvedCode);
                }
                return { body: parse5_1.serialize(documentAst) };
            }
        },
    };
}
exports.babelTransformPlugin = babelTransformPlugin;
